diff -ruN -x '*.orig' -x '*.rej' a/dwm.c b/dwm.c
--- a/dwm.c	2026-09-03 18:34:52.479600285 -0400
+++ b/dwm.c	2026-09-03 18:34:52.481200972 -0400
@@ -86,6 +86,7 @@
 struct Client {
 	char name[256];
 	float mina, maxa;
+	float cfact;
 	int x, y, w, h;
 	int oldx, oldy, oldw, oldh;
 	int basew, baseh, incw, inch, maxw, maxh, minw, minh, hintsvalid;
@@ -208,6 +209,7 @@
 static void setfocus(Client *c);
 static void setfullscreen(Client *c, int fullscreen);
 static void setlayout(const Arg *arg);
+static void setcfact(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setup(void);
 static void seturgent(Client *c, int urg);
@@ -1101,6 +1103,7 @@
 	c->w = c->oldw = wa->width;
 	c->h = c->oldh = wa->height;
 	c->oldbw = wa->border_width;
+	c->cfact = 1.0;
 
 	updatetitle(c);
 	if (XGetTransientForHint(dpy, w, &trans) && (t = wintoclient(trans))) {
@@ -1586,6 +1589,23 @@
 		drawbar(selmon);
 }
 
+void setcfact(const Arg *arg) {
+	float f;
+	Client *c;
+
+	c = selmon->sel;
+
+	if(!arg || !c || !selmon->lt[selmon->sellt]->arrange)
+		return;
+	f = arg->f + c->cfact;
+	if(arg->f == 0.0)
+		f = 1.0;
+	else if(f < 0.25 || f > 4.0)
+		return;
+	c->cfact = f;
+	arrange(selmon);
+}
+
 /* arg > 1.0 will set mfact absolutely */
 void
 setmfact(const Arg *arg)
diff -ruN -x '*.orig' -x '*.rej' a/vanitygaps.c b/vanitygaps.c
--- a/vanitygaps.c	2026-09-03 18:34:52.479708109 -0400
+++ b/vanitygaps.c	2026-09-03 18:34:52.491280251 -0400
@@ -776,8 +776,7 @@
 	int oh, ov, ih, iv;
 	int mx = 0, my = 0, mh = 0, mw = 0;
 	int sx = 0, sy = 0, sh = 0, sw = 0;
-	float mfacts, sfacts;
-	int mrest, srest;
+	float mfacts = 0, sfacts = 0;
 	Client *c;
 
 	getgaps(m, &oh, &ov, &ih, &iv, &n);
@@ -790,20 +789,34 @@
 	sh = m->wh - 2*oh - ih * (n - m->nmaster - 1);
 	sw = mw = m->ww - 2*ov;
 
+	for (i = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+		if (i < m->nmaster)
+			mfacts += c->cfact;
+		else
+			sfacts += c->cfact;
+
 	if (m->nmaster && n > m->nmaster) {
 		sw = (mw - iv) * (1 - m->mfact);
 		mw = mw - iv - sw;
 		sx = mx + mw + iv;
 	}
 
-	getfacts(m, mh, sh, &mfacts, &sfacts, &mrest, &srest);
-
 	for (i = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
 		if (i < m->nmaster) {
-			resize(c, mx, my, mw - (2*c->bw), (mh / mfacts) + (i < mrest ? 1 : 0) - (2*c->bw), 0);
+			if (i == MIN(n, m->nmaster) - 1)
+				mh = m->wh - oh - my; /* last master window takes the rest */
+			else
+				mh = (m->wh - 2*oh - (my - m->wy - oh)) * (c->cfact / mfacts) - ih;
+			resize(c, mx, my, mw - (2*c->bw), mh - (2*c->bw), 0);
 			my += HEIGHT(c) + ih;
+			mfacts -= c->cfact;
 		} else {
-			resize(c, sx, sy, sw - (2*c->bw), (sh / sfacts) + ((i - m->nmaster) < srest ? 1 : 0) - (2*c->bw), 0);
+			if (i == n - 1)
+				sh = m->wh - oh - sy; /* last stack window takes the rest */
+			else
+				sh = (m->wh - 2*oh - (sy - m->wy - oh)) * (c->cfact / sfacts) - ih;
+			resize(c, sx, sy, sw - (2*c->bw), sh - (2*c->bw), 0);
 			sy += HEIGHT(c) + ih;
+			sfacts -= c->cfact;
 		}
 }
\ No newline at end of file
