--- a/drw.c
+++ b/drw.c
@@ -1,5 +1,6 @@
 /* See LICENSE file for copyright and license details. */
 #include <stdio.h>
+#include <X11/Xcursor/Xcursor.h>
 #include <stdlib.h>
 #include <string.h>
 #include <X11/Xlib.h>
@@ -432,7 +433,9 @@
 	if (!drw || !(cur = ecalloc(1, sizeof(Cur))))
 		return NULL;
 
-	cur->cursor = XCreateFontCursor(drw->dpy, shape);
+	cur->cursor = XcursorShapeLoadCursor(drw->dpy, shape);
+	if (!cur->cursor)
+		cur->cursor = XCreateFontCursor(drw->dpy, shape);
 
 	return cur;
 }
